File: friday-sdk-js/src/friday.ts
interface FridayClientConfig {
  apiKey: string;
  baseUrl?: string;
}

interface SearchOptions {
  location?: string;
  numResults?: number;
}

interface CrawlOptions {
  formats?: string[];
  maxPages?: number;
}

interface ScrapeOptions {
  formats?: string[];
}

interface ApiResponse<T = any> {
  data: T;
  error?: string;
}

export class FridayClient {
  private apiKey: string;
  private baseUrl: string;

  /**
   * Initialize the Friday API client.
   * @param apiKey - Your Friday API key
   * @param baseUrl - Optional base URL for the API
   */
  constructor(apiKey: string, baseUrl?: string) {
    this.apiKey = apiKey;
    this.baseUrl = baseUrl?.replace(/\/$/, '') || 'https://api.fridaydata.tech';
  }

  private async makeRequest<T>(
    method: string,
    endpoint: string,
    options: RequestInit = {}
  ): Promise<T> {
    const url = `${this.baseUrl}${endpoint}`;
    const headers = {
      'X-API-Key': this.apiKey,
      'Content-Type': 'application/json',
      ...options.headers,
    };

    try {
      const response = await fetch(url, {
        method,
        headers,
        ...options,
      });

      if (!response.ok) {
        const error = await response.json().catch(() => ({ message: 'Unknown error' }));
        throw new Error(error.message || `HTTP error! status: ${response.status}`);
      }

      return response.json();
    } catch (error) {
      if (error instanceof Error) {
        throw new Error(`Request failed: ${error.message}`);
      }
      throw error;
    }
  }

  /**
   * Fetch and analyze a LinkedIn profile.
   * @param profileUrl - LinkedIn profile URL to analyze
   */
  async getProfile(profileUrl: string): Promise<ApiResponse> {
    return this.makeRequest('GET', `/profile?profile_url=${profileUrl}`);
  }

  /**
   * Analyze a company based on its LinkedIn URL.
   * @param linkedinUrl - Company's LinkedIn URL
   */
  async analyzeCompany(linkedinUrl: string): Promise<ApiResponse> {
    return this.makeRequest('POST', '/analyze-company', {
      body: JSON.stringify({ linkedin_url: linkedinUrl }),
    });
  }

  /**
   * Scrape a website.
   * @param url - URL to scrape
   * @param formats - Output formats (default: ['html'])
   */
  async scrape(url: string, formats: string[] = ['html']): Promise<ApiResponse> {
    return this.makeRequest('POST', '/scrape', {
      body: JSON.stringify({
        url,
        formats,
      }),
    });
  }

  /**
   * Crawl a website.
   * @param url - Starting URL to crawl
   * @param formats - Output formats (default: ['html'])
   * @param maxPages - Maximum number of pages to crawl (default: 10)
   */
  async crawl(
    url: string,
    formats: string[] = ['html'],
    maxPages: number = 10
  ): Promise<ApiResponse> {
    return this.makeRequest('POST', '/crawl', {
      body: JSON.stringify({
        url,
        formats,
        max_pages: maxPages,
      }),
    });
  }

  /**
   * Perform a Google search.
   * @param query - Search query
   * @param location - Search location (default: 'US')
   * @param numResults - Number of results to return (default: 15)
   * @returns Promise<ApiResponse> - Returns either an object or array response
   */
  async search(
    query: string,
    location: string = 'US',
    numResults: number = 5
  ): Promise<ApiResponse> {
    return this.makeRequest('POST', '/search', {
      body: JSON.stringify({
        query,
        location,
        num_results: numResults,
      }),
    });
  }

  /**
   * Extract specific information from a website using AI.
   * @param url - Website URL to analyze
   * @param query - Query describing what information to extract
   */
  async extract(url: string, query: string): Promise<ApiResponse> {
    return this.makeRequest('POST', '/extract', {
      body: JSON.stringify({ url, query }),
    });
  }

  /**
   * Get current API key status and rate limit information.
   */
  async getStatus(): Promise<ApiResponse> {
    return this.makeRequest('GET', '/status');
  }
}

File: friday-sdk-python/friday_sdk/friday.py
import requests
from typing import List, Dict, Optional, Union, Any
from urllib.parse import urljoin

class FridayClient:
    """
    Python SDK for the Friday API.
    
    Args:
        api_key (str): Your Friday API key
        base_url (str, optional): Base URL for the API. Defaults to production URL.
    """
    
    def __init__(self, api_key: str, base_url: str = "https://api.fridaydata.tech/"):
        self.api_key = api_key
        self.base_url = base_url.rstrip('/')
        self.session = requests.Session()
        self.session.headers.update({
            "X-API-Key": api_key,
            "Content-Type": "application/json"
        })

    def _make_request(self, method: str, endpoint: str, **kwargs) -> dict:
        """Make a request to the API."""
        url = urljoin(self.base_url, endpoint.lstrip('/'))
        response = self.session.request(method, url, **kwargs)
        response.raise_for_status()
        return response.json()

    def get_profile(self, profile_url: str) -> dict:
        """
        Fetch and analyze a LinkedIn profile.
        
        Args:
            profile_url (str): LinkedIn profile URL to analyze
            
        Returns:
            dict: Analyzed profile data
        """
        return self._make_request("GET", "/profile", params={"profile_url": profile_url})

    def analyze_company(self, linkedin_url: str) -> dict:
        """
        Analyze a company based on its LinkedIn URL.
        
        Args:
            linkedin_url (str): Company's LinkedIn URL
            
        Returns:
            dict: Company analysis data
        """
        return self._make_request("POST", "/analyze-company", json={"linkedin_url": linkedin_url})

    def scrape(self, url: str, formats: List[str] = ["html"]) -> dict:
        """
        Scrape a website.
        
        Args:
            url (str): URL to scrape
            formats (list): List of formats to return. Options: ["html", "markdown", "links"]
            
        Returns:
            dict: Scraped data in requested formats
        """
        return self._make_request("POST", "/scrape", json={
            "url": url,
            "formats": formats
        })

    def crawl(self, url: str, formats: List[str] = ["html"], max_pages: int = 10) -> dict:
        """
        Crawl a website.
        
        Args:
            url (str): Starting URL to crawl
            formats (list): List of formats to return. Options: ["html", "markdown", "links"]
            max_pages (int): Maximum number of pages to crawl
            
        Returns:
            dict: Crawled data for each page
            
        Raises:
            requests.exceptions.HTTPError: If the API request fails
            ValueError: If invalid formats are provided
        """
        try:
            return self._make_request("POST", "/crawl", json={
                "url": url,
                "formats": formats,
                "max_pages": max_pages
            })
        except requests.exceptions.HTTPError as e:
            raise e

    def search(self, query: str, location: str = "US", num_results: int = 15) -> dict:
        """
        Perform a Google search.
        
        Args:
            query (str): Search query
            location (str): Country code for search location (e.g., "US", "UK")
            num_results (int): Number of results to return
            
        Returns:
            dict: Search results including titles, URLs, and snippets
        """
        return self._make_request("POST", "/search", json={
            "query": query,
            "location": location,
            "num_results": num_results
        })

    def extract(self, url: str, query: str, custom_schema: Optional[Union[Dict[str, Any], List[Dict[str, Any]]]] = None) -> dict:
        """
        Extract specific information from a website using AI.
        
        Args:
            url (str): Website URL to analyze
            query (str): Query describing what information to extract
            custom_schema (Union[Dict[str, Any], List[Dict[str, Any]]], optional): Custom schema for structuring the extracted data
            
        Returns:
            dict: Extracted information
        """
        payload = {
            "url": url,
            "query": query
        }
        if custom_schema is not None:
            payload["custom_schema"] = custom_schema
            
        return self._make_request("POST", "/extract", json=payload)